File: math/minimum-time-difference.cpp
/*
  Given a list of 24-hour clock time points in "HH:MM" format, return the minimum minutes difference between any two time-points in the list.
  
  Example 1:
    Input: timePoints = ["23:59","00:00"]
    Output: 1

  Example 2:
    Input: timePoints = ["00:00","23:59","00:00"]
    Output: 0
*/
class Solution {
public:
    int findMinDifference(vector<string>& timePoints) {
        vector<bool> times(1440);
        for (string& time : timePoints) {
            int minutes = to_minutes(time);
            if (times[minutes]) return 0;
            times[minutes] = true;
        }

        int prev = -1, first = -1, next = 0;
        int m = INT_MAX;
        for (size_t i = 0; i < 1440; ++i) {
            if (times[i]) {
                if (prev < 0) first = i;
                else {
                    next = i;
                    m = min(m, next - prev);
                }
                prev = i;
            }
        }

        return min(m, first - prev + 1440);
    }
private:
    static int to_minutes(string& time) {
        int hour = stoi(time.substr(0, 2));
        int minute = stoi(time.substr(3, 6));

        return hour * 60 + minute;
    }
};

File: math/minimum-time-difference.rs
/*
  Given a list of 24-hour clock time points in "HH:MM" format, return the minimum minutes difference between any two time-points in the list.
  
  Example 1:
    Input: timePoints = ["23:59","00:00"]
    Output: 1

  Example 2:
    Input: timePoints = ["00:00","23:59","00:00"]
    Output: 0
*/
impl Solution {
    pub fn find_min_difference(time_points: Vec<String>) -> i32 {
        let mut times = [false; 1440];

        for time in time_points {
            let minutes = Self::to_minutes(&time) as usize;
            if times[minutes] { return 0; }
            times[minutes] = true;
        }

        let mut prev = -1;
        let mut first = -1;
        let mut next = 0;
        let mut min = i32::MAX;

        for i in 0..1440 {
            if times[i] {
                if prev < 0 { 
                    first = i as i32;
                } else {
                    next = i as i32;
                    min = min.min(next - prev);
                }
                prev = i as i32;
            }
        }

        min.min(first - prev + 1440)
    }

    fn to_minutes(time: &str) -> i32 {
        let hour: i32 = time[0..2].parse().unwrap();
        let minute: i32 = time[3..].parse().unwrap();

        hour * 60 + minute
    }
}
